packager/packager.cc

```diff
@@ -598,35 +603,64 @@ class PackageInfoImpl final : public core::packages::PackageInfo {
         }
     }
 
-    bool importsTransitively(const core::GlobalState &gs, const core::packages::MangledName &otherPkg) const {
-        UnorderedSet<core::packages::MangledName> seen;
-        vector<core::packages::MangledName> toVisit;
-        toVisit.push_back(mangledName());
+    string renderPath(const core::GlobalState &gs, vector<core::packages::MangledName> path) const {
+        // TODO(neil): if the cycle has a large number of nodes (10?), show partial path (first 5, ... (n omitted), last
+        // 5) to prevent error being too long
+        std::string pathMessage;
+        for (int i = 0; i < path.size(); i++) {
+            auto name = gs.packageDB().getPackageInfo(path[i]).show(gs);
+            bool showArrow = i < path.size() - 1;
+            pathMessage += core::ErrorColors::format("    `{}`{}\n", name, showArrow ? " →" : "");
+        }
+        return pathMessage;
+    }
+
+    optional<string> pathTo(const core::GlobalState &gs, const core::packages::MangledName dest) const {
+        auto src = mangledName();
+        queue<core::packages::MangledName> q;
+        // Maps from package to what package we came from to get to that package, used to construct the path
+        // Ex. A -> B -> C means that prev[C] = B and prev[B] = A
+        UnorderedMap<core::packages::MangledName, core::packages::MangledName> prev;
+        UnorderedSet<core::packages::MangledName> visited;
 
-        while (!toVisit.empty()) {
-            auto current = toVisit.back();
-            toVisit.pop_back();
-            if (seen.contains(current)) {
+        q.push(src);
+        while (!q.empty()) {
+            auto curr = q.front();
+            q.pop();
+            if (visited.contains(curr)) {
                 continue;
             }
-            seen.insert(current);
-
-            if (current == otherPkg) {
-                return true;
+            visited.insert(curr);
```

Comment:
Minor, but I learned this the other day:

https://en.cppreference.com/w/cpp/container/set/emplace

You can use

```cpp
auto [_it, inserted] = visited.emplace(curr);
```

to insert `curr` if it's not there, and still be able to check whether the insert happens (so that you only have to do the hashmap search operation once)